docs/source/config.rst
Configuration
=============

The application is designed with a high degree of flexibility, 
allowing users to tailor its behavior and appearance to suit their needs. 

Configuration is stored in `config.json <configfile.html>`_ which should be located alongside the binary file.
It is read at application launch.

Key configuration options include:

    - **Frames Per Second (FPS)**: Users can define the desired update frequency for rendering and data visualization.

    - **Buffer size**: Defines the number of time-related datapoints are kept in buffer. 
        
        .. note::
            For slower systems this value shoud be increased.

    - **Padding**: Defines the extra space added to graphs when signals are autoscaled. In percentage.
    - **Model configuration**: Select which solver to use.

        - **Exported signals**: Select which signals the model exports to the buffer.
        - **Parameter Adjustments**: Choose which model parameters can be directly modified through the GUI.
    
    - **Customization of Signals**: Configure attributes such as colors and units for signals.

.. toctree::
    :hidden:

    configfile
    datacontainers
    modelconf
    guiconf



docs/source/configfile.rst
`config.json`
=============

The configuration file for the application defines various settings and parameters. Below is the structure of the configuration file, along with examples.

Top-Level Structure
-------------------

- **ModelWrapper**: Defines simulation-related parameters, including the solver and signals to export.

    - **solver** 
    - **export_signals**: Signals appended to buffer after every solver timestep

        - **name**: Used as a key when specifying which signals are to be displayed on a graph. See `PVGraph`, `SSGraph`, `GraphGrid`.
        - **path**: Path to a speficified CircAdapt signal. See CircAdapt documentation for details.
        - **type**: Specifies the type of the signal. Needed for correct conversion of values between the model and UI. There are several predefined types, see :ref:`datacontainers` for more information.

        .. note::
            Some additional signals are exported to buffer. They are not specified
            in the user config file because they are needed for certain calculations.

            See `ModelWrapper::setupAdditionalSignals`

    - **parameters**: Parameters which can be changed from the GUI.

        - **name**: Used as a key at defining GUI elements for parameter modification. See `ModelParameters`.
        - **path**: Path to a speficified CircAdapt parameter. See CircAdapt documentation for details.
        - **type**: Specifies the type of parameter. Needed for correct conversion of values between the model and UI. There are several predefined types, see :ref:`datacontainers` for more information.

- **GUI**: 

    - **framerate**: How many frames are displayed per second.
    - **buffersize**: Number of time-related datapoints kept in buffer.
    - **padding**: Extra space around the signals when using autoscale function. In percent with respect to graph coordinates.
- **PVGraph**: Configures pressure-volume graphs, including titles and signal mappings.

    - **title**

    - **xLabel**
    - **yLabel**

    - **Signals**: Array of signals to be displayed on this graph.

        - **displayName**: Name of the signal, visible in tooltip on hover or right-click menu.

        - **xVar**: Key of signal used for x-value.
        - **yVar**: Key of signal used for y-value.
        - **isVisible**: Is it shown when starting the application.
        - **unit**

        - **color**

- **SSGraph**: Configures stress-strain graphs, including titles and signal mappings.

    - **title**

    - **xLabel**
    - **yLabel**

        - **displayName**: Name of the signal, visible in tooltip on hover or right-click menu.

        - **xVar**: Key of signal used for x-value.
        - **yVar**: Key of signal used for y-value.
        - **isVisible**: Is it shown when starting the application.
        - **unit**

        - **color**

- **GraphGrid**: Defines signals grouped by type. Each type regards one row. Each row has 3 columns: `REFERENCE`, `SNAPSHOT`, `CURRENT`.

    Row types are:

    .. doxygenenum:: GraphGrid::RowType

    In `config.json`, each row specifier takes a list of signals to be displayed in that row. Keys for defining signals are:

        - **displayName**: Name of the signal, visible in tooltip on hover or right-click menu.

        - **name**: Key of the signal corresponding to one of `export_signals`.
        - **isVisible**: Is it shown when starting the application.
        - **unit**
        - **color**

        - **inMainMenu**: Indicates whether the signal is displayed directly in the main menu upon right-click or nested within the `More` submenu.

- **ValueView**: Specifies metrics which are shown in the table (eg. CO, mLAP, ...).
    
    A list of objects with keys:

        - **name**: Corresponds to an element of m_beatStats from Buffer class. See Buffer.
        
        - **unit**



Example Structure
-----------------

.. code-block:: json

    {
        "ModelWrapper": {
            "solver": "forward_euler",
            "export_signals": [
                {
                    "name": "pLa",
                    "path": "Model.Peri.La.p",
                    "type": "pressure"
                },
                {
                    "name": "pRa",
                    "path": "Model.Peri.Ra.p",
                    "type": "pressure"
                },
                ...
            ],
            "parameters": [
                {
                    "name": "Left atrial contractility",
                    "path": "Model.Peri.La.Sf_act",
                    "type": "percentage"
                },
                ...
            ]
        },
        "GUI": {
            "framerate": 60,
            "buffersize": 50,
            "padding": 0.05
        },
        "PVGraph": {
            "title": "Pressure-volume loops",
            "xLabel": "Volume [mL]",
            "yLabel": "Pressure [mmHg]",
            "Signals": [
                {
                    "displayName": "Left atrium",
                    "xVar": "VLa",
                    "yVar": "pLa",
                    "isVisible": false,
                    "color": "#FB9A99"
                },
                ...
            ]
        },
        "SSGraph": {
            "title": "Stress-strain loops",
            "xLabel": "Strain [%]",
            "yLabel": "Stress [kPa]",
            "Signals": [
                {
                    "displayName": "Left atrium",
                    "xVar": "strain_La",
                    "yVar": "Sf_La",
                    "isVisible": false,
                    "color": "#FB9A99"
                },
                ...
            ]
        },
        "GraphGrid": {
            "PRESSURE": [
                {
                    "name": "pLa",
                    "displayName": "Left atrium",
                    "isVisible": false,
                    "unit": "mmHg",
                    "color": "#FB9A99",
                    "inMainMenu": true
                },
                ...
            ],
            "VOLUME": [
                ...
            ],
            ...
        },
        "ValueView": [
            {
                "name": "CO",
                "unit": "L/min"
            },
            {
                "name": "mLAP",
                "unit": "mmHg"
            },
            ...
        ]
    }


CircAdaptUI/settings.h
#ifndef SETTINGS_H
#define SETTINGS_H

#include <QObject>
#include <QSettings>
#include <QJsonObject>
#include <QJsonDocument>
#include <QJsonArray>

class Settings : public QObject {
    Q_OBJECT
public:
    static Settings& instance() {
        static Settings instance;
        return instance;
    }

    void load(QString filepath);
    QJsonObject GraphGrid() const {  return jsonDoc["GraphGrid"].toObject(); };
    QJsonArray ExportSignals() const {  return jsonDoc["ModelWrapper"]["export_signals"].toArray(); };
    QJsonArray ExportParameters() const {  return jsonDoc["ModelWrapper"]["parameters"].toArray(); };
    QJsonArray ValueView() const {  return jsonDoc["ValueView"].toArray(); };
    QJsonObject PVGraph() const {  return jsonDoc["PVGraph"].toObject(); };
    QJsonObject SSGraph() const {  return jsonDoc["SSGraph"].toObject(); };
    QJsonObject MWSettings() const { return jsonDoc["ModelWrapper"].toObject(); };
    QJsonObject ModelParameters() const { return jsonDoc["ModelParameters"].toObject(); };

    int fps() const {
        return m_settings.value("GUI/framerate", 2).toInt();
    }

    int beatIdx() const {
        return m_settings.value("ModelWrapper/beatIdx", 0).toInt();
    }

    int buffersize() const {
        return m_settings.value("buffersize", 60).toInt();
    }

    double paddingVertical() const {
        return m_settings.value("GUI/padding", 0.05).toDouble();
    }

    void beatIdxIncrement() {
        m_settings.setValue("ModelWrapper/beatIdx", beatIdx() + 1);
    }

signals:
    void fpsChanged(int fps);
public slots:

    void setFps(int fps) {
        m_settings.setValue("framerate", fps);
        emit fpsChanged(fps);
    }
    void setBeatIdx(int idx) {
        m_settings.setValue("ModelWrapper/beatIdx", idx);
    }

private:
    Settings() :
        m_settings("CircAdapt", "Andrija") {};
    ~Settings() {};

    void parseJsonObject(const QJsonObject &jsonObj, const QString &prefix = "");
    void setDefaultConfig();
    QJsonDocument jsonDoc{};

    Settings(const Settings&) = delete;
    Settings& operator=(const Settings&) = delete;

    QSettings m_settings;
};


#endif // SETTINGS_H
